lib/Smr/Container/DiContainer.php

```diff
@@ -0,0 +1,112 @@
+<?php declare(strict_types=1);
+
+namespace Smr\Container;
+
+use DI\Container;
+use DI\ContainerBuilder;
+use Dotenv\Dotenv;
+use MySqlDatabase;
+use mysqli;
+use RuntimeException;
+use Smr\MySqlProperties;
+use function DI\autowire;
+
+/**
+ * Class DiContainer
+ * A wrapper around the DI\Container functionality that will allow
+ * static usage of container methods.
+ * @package Smr\Container
+ */
+class DiContainer {
+	private static DiContainer $instance;
+	private Container $container;
+
+	private function __construct() {
+		self::$instance = $this;
+		$this->container = $this->buildContainer();
+	}
+
+	private function getDefinitions(): array {
+		return [
+			/*
+			 * mysqli is a 3rd party library, and we do not have control over its constructor.
+			 * In order for PHP-DI to construct an instance of a class, each constructor argument must
+			 * be able to be constructed.
+			 * PHP-DI cannot construct mysqli by itself, because all of its arguments are primitive types.
+			 * Therefore, we need to declare a provider factory for the container to use when constructing new instances.
+			 *
+			 * The factories themselves are able to use dependency injection as well, so we can provide the MysqlProperties
+			 * typehint to make sure the container constructs and instance and provides it to the factory.
+			 */
+			mysqli::class => function (MySqlProperties $mysqlProperties): mysqli {
+				// Set the mysqli driver to raise exceptions on errors
+				if (!mysqli_report(MYSQLI_REPORT_ERROR | MYSQLI_REPORT_STRICT)) {
+					throw new RuntimeException('Failed to enable mysqli error reporting');
+				}
+				return new mysqli(
+					$mysqlProperties->getHost(),
+					$mysqlProperties->getUser(),
+					$mysqlProperties->getPassword(),
+					$mysqlProperties->getDatabaseName());
+			},
+			Dotenv::class => function (): Dotenv {
+				return Dotenv::createArrayBacked(ROOT);
+			},
+			// Explicitly name all classes that are autowired, so we can take advantage of
```

Comment:
I'm a bit confused about these class enumerations. One of the benefits of autoloader is that you only load the files you need. If we use DI throughout the entire library, and we have to explicitly specify them here, then that means we'll be loading all the files as soon as we load the `DiContainer`. (And I suspect this wouldn't work for classes that use the game-type dependent loading, e.g. `lib/Default` vs. `lib/Draft`, which is kind of a hack, so maybe this will open the door for a better design for that.)

Now, I do think that we eventually want to take advantage of preloading in at least the production image (already available in PHP 7.4). That might make my concern about overzealous autoloading moot.

Reply:
You don't have to specify them, I just added them explicitly so that the definitions can be compiled. Also, everything that I've defined in here are already loaded in every page request -- the `MySqlDatabase` is constantly being used everywhere. So we're not loading anything that doesn't need to be.

Reply:
Also about the game depend loading -- let's cross that bridge when we get there. There are several more iterations to go through before this thing starts to jump out at you and provide an "aha" moment.

Reply:
Okay, sounds good.